### ISSUE TITLE
Compound (multi-value) inputs?

### ISSUE THREAD

**Issue by mootari** [observablehq/inputs#73]
Something that I find myself doing again and again is to join multiple inputs into a single value object. The value object acts as a proxy to the individual inputs.

It would be great to have such a helper as part of the library.

Example ([demo](https://observablehq.com/d/f3bc3f9ea45b9f73)):
```javascript
function mergeInputs(inputs, target = {}) {
  const props = Object.fromEntries(Object.entries(inputs).map(([key, input]) => [key, {
    enumerable: true,
    configurable: true,
    get: () => input.value,
    set: value => input.value = value,
  }]));
  return Object.defineProperties(target, props);
}
```
```javascript
viewof opts = {
  const inputs = {
    n: Range([1, 30], {value: 19, step: 1, label: "size" }),
    a: Range([0, 45], {value: 27.5, step:.1, label: "angle" }),
    x: Range([-100, 100], {value: 0, step: 1, label: "x offset" }),
    y: Range([-100, 100], {value: 0, step: 1, label: "y offset" }),
  };
  const view = html`<div>${Object.values(inputs)}`;
  view.value = mergeInputs(inputs);
  return view;
}
```
```
// opts
Object {n: 19, a: 27.5, x: 0, y: 0}
```

**Comment by mbostock** [comment]
Thanks for the suggestion! I’d definitely like us to have a recommended approach for “compound” inputs, or denser layout of multiple inputs.

Here’s an alternative I’ve been considering. It doesn’t require any new machinery, and it lets you reference individual values reactively. The downside is you end up with some orphaned cells because the views are ripped out of their original cell and then inserted into the combining cell.

```js
html`${[viewof n, viewof a, viewof x, viewof y]}`
```
```js
viewof n = Range([1, 30], {value: 19, step: 1, label: "size" })
```
```js
viewof a = Range([0, 45], {value: 27.5, step: 0.1, label: "angle" })
```
```js
viewof x = Range([-100, 100], {value: 0, step: 1, label: "x offset" })
```
```js
viewof y = Range([-100, 100], {value: 0, step: 1, label: "y offset" })
```

It should look like this:

<img width="1020" alt="Screen Shot 2021-02-07 at 10 14 11 AM" src="https://user-images.githubusercontent.com/230541/107155443-3e9f3c80-692d-11eb-8f9a-8e6c6ebf3a22.png">

**Comment by mootari** [comment]
I see the primary application for inputs which only make sense when used together (that is, you would combine them into an object or array anyway). In my opinion, having to splitting them up across multiple cells creates unnecessary mental overhead.

I also try to stay clear of reparenting, because I think it's confusing to viewers. At the very least the moved element could be replaced with a notice, e.g.:
```javascript
function move(target) { return target.replaceWith(html`<em>This element has been moved.`), target }
```
```javascript
md`Adjust here: ${move(viewof foo)}`
```
Other benefits of compound inputs: You can treat the input library (or really any view) as widgets and generate combinations of them dynamically (.e.g. options based on previous choices, array inputs, compound row inputs etc); inputs can be moved around as one cell, which makes a significant difference for large numbers of inputs.

**tl;dr:** I don't think reparenting is a suitable alternative. What would be your expectations and requirements for a compound widget or function, to accept it into the library?

**Comment by j-f1** [comment]
To resolve the issues with reparenting, the Observable runtime could detect when the topmost DOM node is removed and replace it with a standard `viewof x = HTMLElement { ... }` inspector so there isn’t a blank spot left.

**Comment by mbostock** [comment]
Nice idea, @j-f1. That would require a shallow mutation observer for every cell, but I think that’s doable.

That said, I do think it’s worthwhile to have an explicit construct for this, rather than the “one weird trick” of defining a DOM element in one cell and “stealing” it from another. If only because it means you only need one cell to declare your inputs. If we eventually support cell destructuring (https://talk.observablehq.com/t/feature-request-destructuring-assignment/1127) you get the best of both worlds.

If I knew what my requirements were, I would just build the thing. 😄 For now I’m interested in proposals, examples, and debate on various approaches.

**Comment by geofduf** [comment]
Maybe having a built in function similar to https://observablehq.com/@mbostock/form-input would make sense ?

**Comment by mootari** [comment]
> Maybe having a built in function similar to https://observablehq.com/@mbostock/form-input would make sense ?

This library's widgets are wrapped in individual form elements, creating two challenges:
- Form elements cannot be nested, and relying on name attributes may no longer work in this context.
- The form element acts as the *actual* input. Anything inside it should be considered internal.

The value object has to be constructed and maintained regardless, so it makes sense to start with a low-level function first. This function will then allow you to create opinionated solutions that rely on certain conventions.

While I was working on a spec, some questions came up:
- How should values get assigned? Am I allowed to reassign individual properties, or am I required to assign the object as a whole (similar to `mutable` assignments)?
- Should the value be detached from the inputs? I.e., would each change create a new object and set it as value of the compound view?

**Comment by tomlarkworthy** [comment]
These examples are the cross product of n *independent* controls. Real composite controls are more complex with mixed concerns. Then it is much harder to write the "set" function and to avoid self triggering event loops. 

I would argue that the fact you can decompose the function into distinct cells shows the composite is factorable, and maybe distinct cells is the right representation (except that it takes up too much vertical space and the DX is not so good). Factorable UIs are the *easy* case though, and not the case where expressivity is amplified. 

I good composite control is like Unity with a treeview scene graph on the left and a 3d view in the middle. Both are projections of a common world, but one is optimised for spatial reasoning, one for list reasoning. The combination is strictly more powerful than either alone (e.g. selection is shared).

@mootari suggestion on https://observablehq.com/@tomlarkworthy/audio-inputs 
"Alternatively: can start and length of the selection be hooked up to knobs?"
is an exact example of what is not covered here, but what should be covered here. What if "x" is constrained to be the sqrt of "a"

Also "@observablehq/inputs" provide examples of not factorable composite inputs (https://github.com/observablehq/inputs/blob/main/src/range.js#L42). For a Range, 1. if the number is updated change the slider. 2. If the slider updates update the number. I think the fact you can't replicate any of the controls in "@observablehq/inputs" is why this is not solving a hard enough problem yet.

**Comment by mootari** [comment]
As far as I understand, the inputs combine both model and view. What you're suggesting sounds more like a separate model with multiple views attached (which can be done e.g. through mutable). However, I have to wonder at what point we're just emulating the reactivity of the Runtime inside a cell.

**Comment by tomlarkworthy** [comment]
I do think allowing "value" to be writable is a good standard which allows other code to reuse controls better. It's a mechanism in mergeInputs as part of this proposal but I think in general it's something most controls should do. 

With writability a control becomes a read/write view not a read-only view. Given most viewof "own" the storage of the source of truth, making them mutable gets them closer to being a general purpose variable that happens to have a UI attached.

**Comment by mbostock** [comment]
Here’s a variation of @mootari’s proposal that implements value assignment and also allows you to define an arbitrary structure for the compound value:

```js
viewof options = {
  const n = Range([1, 30], {value: 19, step: 1, label: "size"});
  const a = Range([0, 45], {value: 27.5, step: 0.1, label: "angle"});
  const x = Range([-100, 100], {value: 0, step: 1, label: "x offset"});
  const y = Range([-100, 100], {value: 0, step: 1, label: "y offset"});
  return Object.defineProperty(html`${[n, a, x, y]}`, "value", {
    get() {
      return {
        n: n.value,
        a: a.value,
        p: [x.value, y.value]
      };
    },
    set({n: _n, a: _a, p: _p}) {
      if (_n !== undefined) n.value = _n;
      if (_a !== undefined) a.value = _a;
      if (_p !== undefined) [x.value, y.value] = _p;
    }
  });
}
```

One thing missing from the above implementation is that it’d be better if the value object were the same if the value hasn’t changed; currently `viewof options.value === viewof options.value` is false because the getter creates a new object on each get. It’s better to cache the last value and then clear it when there’s an *input* event. Another missing thing is that if any of the child inputs don’t emit *input* events, e.g. bare buttons, then it won’t work because Generators.input only listens for *input* events on the compound input. But maybe that’s fine, especially if this is intended primarily to compose other Observable Inputs.

I feel shorthand (convenience) is likely the main draw here, so here’s a more constrained take:

```js
function Form(inputs) {
  let value; // lazily constructed
  const root = html`<div oninput=${() => value = undefined}>${Object.values(inputs)}`;
  return Object.defineProperty(root, "value", {
    get() {
      return value === undefined
        ? (value = Object.fromEntries(Object.entries(inputs).map(([name, input]) => [name, input.value])))
        : value;
    },
    set(v = {}) {
      value = undefined;
      for (const name in v) {
        if (name in inputs && v[name] !== undefined) {
          inputs[name].value = v[name];
        }
      }
    }
  });
}
```

Then you’d say:

```js
viewof options = Form({
  n: Range([1, 30], {value: 19, step: 1, label: "size"}),
  a: Range([0, 45], {value: 27.5, step: 0.1, label: "angle"}),
  x: Range([-100, 100], {value: 0, step: 1, label: "x offset"}),
  y: Range([-100, 100], {value: 0, step: 1, label: "y offset"})
})
```

Live example: https://observablehq.com/d/0fa0562e1dbf9542

The idea being: the display is always a simple concatentation of the inputs, and the structure is always a flat object. (We could be a bit more flexible here with options, e.g. allowing a grid, or allowing the value to be an array instead of an object, and possibly even a nested structure.)

**Comment by mbostock** [comment]
Oh, another thing I’d want here is an optional Submit button so that you can adjust the child inputs and then commit the value by clicking submit. I think that should be possible without much difficulty for the Form function (just tack a Submit button onto the end).

**Comment by tomlarkworthy** [comment]
```
"it’d be better if the value object were the same if the value hasn’t changed"
```

This seems like a "design guidline", similar to how a views value should be mutable.

I collected them here: https://observablehq.com/d/8cf099730536c682

Note that cacheing the value is kinda tricky to square with letting events bubble.

I am also thinking that the ui "viewof" should communicate with "backend" value via state transfer *not* action driven function calls.

So instead of having a play() function that is called by the onclick of a button, you create a isPlaying state variable, and toggle that by writing to it.
~~~js
onclick= () => {value.isPlaying = true} 
~~~

This makes it simpler for downstream composite controls to automate their upstream dependancies. They can read state in the same way the write state. It's the same argument for *resource orientated API design* and can be made more robust against self triggering loops etc as it's an idempotent interface.

**Comment by tomlarkworthy** [comment]
I am not sure if this is a bug or not, but when using the "set", an input event is not raised, so I can mutate the control so the UI changes, but any models listening to the UI don't hear and therefore don't update. 

> Here’s a variation of @mootari’s proposal that implements value assignment and also allows you to define an arbitrary structure for the compound value:
> 
> ```js
> viewof options = {
>   const n = Range([1, 30], {value: 19, step: 1, label: "size"});
>   const a = Range([0, 45], {value: 27.5, step: 0.1, label: "angle"});
>   const x = Range([-100, 100], {value: 0, step: 1, label: "x offset"});
>   const y = Range([-100, 100], {value: 0, step: 1, label: "y offset"});
>   return Object.defineProperty(html`${[n, a, x, y]}`, "value", {
>     get() {
>       return {
>         n: n.value,
>         a: a.value,
>         p: [x.value, y.value]
>       };
>     },
>     set({n: _n, a: _a, p: _p}) {
>       if (_n !== undefined) n.value = _n;
>       if (_a !== undefined) a.value = _a;
>       if (_p !== undefined) [x.value, y.value] = _p;
>     }
>   });
> }
> ```


Q: Should setting a view fire an input event?

**Comment by mbostock** [comment]
> I am not sure if this is a bug or not, but when using the "set", an input event is not raised

This is intentional and matches the behavior of native inputs. You have to dispatch an event explicitly if desired.

**Comment by geofduf** [comment]
> (We could be a bit more flexible here with options, e.g. allowing a grid, or allowing the value to be an array instead of an object, and possibly even a nested structure.)

Could it lead to the support of arbitrary layouts in the future ? Or should this still be considered a totally distinct concern ?

Just an example ([live](https://observablehq.com/d/a55e7f007d977547)) :

```javascript
function Form(layout, formatted = false) {
  let value; // lazily constructed
  const root = html`<div oninput=${() => value = undefined}>${formatted ? layout : Object.values(layout)}`;
  const inputs = formatted ? parse(layout) : layout;
  return Object.defineProperty(root, "value", {
    // ...
  });
}
```

**Comment by tomlarkworthy** [comment]
> This is intentional 

Good to know! I added it.

Another thing I noticed, is that for controls wrapping primitive values, the writable property is defined on the ui container. So to actually mutate a value you would call it like:-

```js
    viewof myView.value = "newVal"
```

But for mergeInputs, we are one level deeper, which is simpler to user

```
myview.value.field = "newFieldValue"
```

So It be nice syntactically nicer if we could drop the "viewof" and assign the value directly:

```
myView = "newValue"
```

But we get Syntax error for assigning to constant. Also we would be self trigger, so with that in mind, I kinda think the following should be possible

```
mutable myView = "newValue"
```

Anyway, thats a minor thing, my overall question is do we think the root composite object value should be writable? Or just the subfields? Or both?

Personally I think the root *has* to be writable, for symmetry with primitive controls (https://github.com/observablehq/inputs/blob/f72d22b4a8302922d1f86adb15fdfe6289950254/src/range.js#L47). However, thats the least ergonomic to use. If I want to flip a subfield through the setter of a composite I need to do

viewof myView.value = {...viewof myView.value, subfield: "newValue"}

Interestingly @mootari went for mutability on subfields, and @mbostock on the root.

I think a *good* composite needs to support both.

**Comment by harrybiddle** [comment]
I'm trying to do something I think might be relevant to this issue.

I've been trying to use these inputs inline, a-la [@mbostock/inline-inputs](https://observablehq.com/@mbostock/inline-inputs), to achieve something like this:

<img width="659" alt="image" src="https://user-images.githubusercontent.com/5887227/108604068-e8f67580-73ab-11eb-9260-6b54208d6e22.png">

I find this can be a more engaging way of getting input from the user than a grid layout.

Unfortunately, since the input elements in this library come back wrapped in `<form>` they automatically get displayed as block elements. I can strip back the outer wrapping elements until I get the `<input>` and manually set the CSS on it, but it feels like I'm rather hacking the library!

It would be great if the solution for multi-value inputs would support more than just grid layouts.

**Comment by tomlarkworthy** [comment]
I reconsidered supporting both types of mutation and think it only makes sense to support mutability of the root node, as that one is interesting as it sorta connects the two layers of a viewof (the UI bit and the data bit). I tried to write up the least contentious rules here: https://observablehq.com/@tomlarkworthy/ui-linter with some automated testing. 

I am more interested in general reusability than just compound inputs. Maybe I do not have much to contribute on this narrower topic.

**Comment by zawinul** [comment]
Hello everybody.
I’ve realized a parametric form, Panel, which has some of the features you are talking about.
I didn't know about this thread, thanks @mootari for signaling, so some aspect of my solution overlaps with something already discussed, sorry for that.

My Panel
- initializes calling a function with a descriptor where each control has a name, for example:

		viewof ui = Panel ({ 	
			myFirstPar:  Range ([0, 11], {value: 5, label: "first val"}),
			mySecondPar: Range ([0, 1000], {value: 5, label: "another val"}),
			fruit:       Select(['apple','orange','lemon'], {label: "fruit"}),
			myLastPar:   Toggle ({label: "this is a toggle"})
		})  


- uses @observablehq/inputs fields
- all values ​​can be extracted simultaneously through the expression `ui.asObject` which is updated every time a parameter is changed
- each parameter can be extracted individually by name (eg: `ui.myFirstPar`). This value updates only when that parameter is changed
- a subset of parameters can be extracted through the group function (eg: `ui.group ('myFirstPar', 'myLastPar')`). This function returns an object that is updated only when one of the subset parameters is changed.
- the read function allows you to read the parameter values ​​as an object immediately (`ui.read()`)
- the set function (example `ui.set('myFirstPar', 10)`) allows to set a value and triggers updates where needed
- html rendering is automated, some little customization is possible.

You can find it, with some examples, at https://observablehq.com/@paoloandrenacci/panel/2, I would like very much to have some feedback and to know if it has been useful to you.

**Comment by tomlarkworthy** [comment]
My entry into the composite view https://observablehq.com/@tomlarkworthy/view 

This add a [key, DOM] special case to the hypertext literal. So you just write your compound view in HTML and it can assemble/disassemble on get/set . 

```js
viewof composite = view`<div style="display: flex">
<table>
  <tr>
    <td>${["r1", Range([0, 10])]}</td>
  </tr>
  <tr>
    <td>${["r2", Range([0, 3])]}</td>
  </tr>
  <tr>
    <td>${[
      "text",
      Text({
        label: "Enter some text"
      })
    ]}</td>
  </tr>
</table>
<img width="200"src="https://media.giphy.com/media/2vobTwCkFg88ZUnilt/giphy-downsized.gif"></img>
</div>
`
```
